### PR TITLE
AP-9751 # Added submission timestamp conditional logic

### PR DESCRIPTION
**Pull request by mymattcarroll**
## Requester Checklist

Please only check the items that you have actioned. Do not check items that are not applicable to your PR.

### Implementation

- [ ] Have you tested your implementation locally?
- [ ] If applicable, has an appropriate changelog entry been added? Do not include if you are fixing/changing something that is only in the current release.
- [ ] There are no warnings that have been suppressed unnecessarily
- [ ] Have automated tests been added, or have related ones been updated to cover the change?
- [ ] Have all OneBlink dependency updates been completed (eg apps / apps-react / types etc).
- [ ] Have you ensured this change does not add unwanted dependencies?
- [ ] If this PR contains a refactor, have relevant Jira testing tasks added?
- [ ] Changes that will knowingly make the feature/bug incomplete have been commented with TODO and a description of what needs to be done to finish the feature/bug
- [ ] Any changes made to public APIs have been reflected in the documentation
- [ ] Have you isolated business logic where possible to allow for unit testing?

### Logging and Debugging

- [ ] Are the error messages, if any, informative?
- [ ] Are there enough log events and are they written in a way that allows for easy debugging?
- [ ] "Debugging" code removed
- [ ] Front-end: No erroneous `Console.WriteLines`

### Readability

- [ ] All class, variable, property and method modifiers are provided with the smallest scope possible
- [ ] New files, variables and functions are descriptive/comprehensible and named consistently.
- [ ] There is no dead code (unreachable code)
- [ ] There is no usage of [magic numbers](<https://en.wikipedia.org/wiki/Magic_number_(programming)>)
- [ ] There is no commented out code.
- [ ] In hard-to-understand areas, comments exist and describe rationale or reasons for decisions in code

### Security

- [ ] All personal data inputs are checked (for the correct type, length/size, format, and range).
- [ ] No sensitive information is logged or visible in a stacktrace
- [ ] Are authorization and authentication handled correctly?
- [ ] Is (user) input validated, sanitized, and escaped to prevent security attacks such as cross-site scripting or SQL injection?
- [ ] Is data retrieved from external APIs or libraries checked for security issues?
- [ ] Do API endpoints return appropriate status codes

## Reviewer Guide

- It is important that you understand the purpose of the PR.
- You are encouraged to engage with the requestor if you do not understand any of the proposed code changes/additions/deletions.
- Do you, the reviewer, understand what the code does? Do you think a specific expert, like a security expert or a usability expert, should look over the code before it can be accepted?
- Is a framework, API, library, or service used that should not be used? Are there alternatives you could recommend?
- Are there existing hooks/components/functions in the same code base that could be utilised?
- When reviewing tests, attempt to identify missing edge cases that may be relevant to the proposed implementation
- Ensure you check for:
  - Security
  - Scalability
  - Performance
  - Maintainability

---

> [!NOTE]
> **Medium Risk**
> Multiple documented breaking public API changes affect payment/scheduling and form-fill callers; date predicate correctness depends on caller-supplied parsing/timezone helpers.
> 
> **Overview**
> Adds **`SUBMISSION_TIMESTAMP`** conditional predicates so rules can compare the form submission time to a fixed date or a date/datetime field (with optional day offset), using **`AFTER`** / **`BEFORE`** (exclusive) or **`BETWEEN`** (inclusive). Callers must pass **`submissionTimestamp`**, **`parseDate`**, and **`addDaysToDate`** so timezone and calendar math stay under app control.
> 
> **Breaking:** **`evaluateConditionalPredicates`**, **`paymentService.checkForPaymentEvent`**, and **`schedulingService.checkForSchedulingEvent`** now require those date helpers (payment/scheduling use a single options object). **`generateFormElementsConditionallyShown`** returns **`{ formElementsConditionallyShown, isSubmissionEnabled }`**, accepts optional **`enableSubmission`**, and evaluates submission enablement using already-computed element visibility instead of a separate predicate pass.
> 
> Element-based evaluation is split into **`evaluateFormElementConditionalPredicate`** (renamed from **`evaluateConditionalPredicate`**); several internal helpers move to options-object signatures.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 516d1abe839df1f4ebeecf86ccdd10be9b14f062. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/conditionalLogicService/evaluateConditionalSubmissionTimestampPredicate.ts [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+      const value = formElementsCtrl.model?.[formElementWithName.name]
+      return typeof value === 'string' && value ? value : undefined
+    }
+  } else if (formElementsCtrl.parentFormElementsCtrl) {
+    return getElementValue(
+      formElementsCtrl.parentFormElementsCtrl,
+      elementId,
+    )
+  }
+  return undefined
```
[the remark this answers is not in the document]
Comment (mymattcarroll) [reply src/conditionalLogicService/evaluateConditionalSubmissionTimestampPredicate.ts]: This is by design for repeatable set elements. Nested form elements are out of scope for this work. It may be supported in a future ticket.

File: src/conditionalLogicService/evaluateConditionalSubmissionTimestampPredicate.ts [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+  const date = parseDate(dateString)
+  if (Number.isNaN(date.getTime())) {
+    return undefined
+  }
+
+  if (!dateValue.daysOffset) {
+    return date
+  }
+
+  return addDaysToDate(date, dateValue.daysOffset)
```
[the remark this answers is not in the document]
Comment (mymattcarroll) [reply src/conditionalLogicService/evaluateConditionalSubmissionTimestampPredicate.ts:40]: Fixed

File: src/conditionalLogicService/generateFormElementsConditionallyShown.ts [hunk 3]
```diff
@@ -319,22 +349,79 @@ const generateFormElementsConditionallyShownWithParent = ({
...
+      return !!shown && !shown.isHidden
+    } catch (error) {
+      errorCallback?.(error as Error)
+      return false
+    }
+  }
+
+  const isSubmissionEnabled = enableSubmission.requiresAllConditionalPredicates
+    ? enableSubmission.conditionalPredicates.every(predicateFn)
+    : enableSubmission.conditionalPredicates.some(predicateFn)
```
[the remark this answers is not in the document]
Comment (mymattcarroll) [reply src/conditionalLogicService/generateFormElementsConditionallyShown.ts:421]: `enableSubmission` is not typed as `ConditionalPredicate[]`.

File: src/conditionalLogicService/generateFormElementsConditionallyShown.ts [hunk 4]
```diff
@@ -319,22 +349,79 @@ const generateFormElementsConditionallyShownWithParent = ({
...
+}: FormElementsConditionallyShownParameters): FormElementsConditionallyShownResult {
+  const formElementsConditionallyShown =
+    generateFormElementsConditionallyShownWithParent({
+      formElements,
+      submission,
+      errorCallback,
+      parentFormElementsCtrl: undefined,
+    })
+
+  if (!enableSubmission?.conditionalPredicates.length) {
```
[the remark this answers is not in the document]
Comment (mymattcarroll) [reply src/conditionalLogicService/generateFormElementsConditionallyShown.ts:383]: `conditionalPredicates` is a required property. typescript is enough for us to trust the input data.

File: src/conditionalLogicService/evaluateConditionalSubmissionTimestampPredicate.ts [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+      typeCastService.formElements.toNamedElement(formElement)
+    if (formElementWithName) {
+      const value = formElementsCtrl.model?.[formElementWithName.name]
+      return typeof value === 'string' && value ? value : undefined
+    }
+  } else if (formElementsCtrl.parentFormElementsCtrl) {
+    return getElementValue(formElementsCtrl.parentFormElementsCtrl, elementId)
+  }
+  return undefined
+}
```
Comment (divporter) [inline src/conditionalLogicService/evaluateConditionalSubmissionTimestampPredicate.ts]: Do we not have a reusable function for this? Seems pretty generic.

Comment (mymattcarroll) [reply src/conditionalLogicService/evaluateConditionalSubmissionTimestampPredicate.ts]: Ill have a look

File: src/conditionalLogicService/evaluateConditionalSubmissionTimestampPredicate.ts [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+  )
+  if (!compareDate) {
+    return false
+  }
+
+  if (predicate.operator === 'AFTER') {
+    return submissionTime > compareDate.getTime()
+  }
+
+  return submissionTime < compareDate.getTime()
```
Comment (divporter) [inline src/conditionalLogicService/evaluateConditionalSubmissionTimestampPredicate.ts]: I know it's the only remaining option but I'd be more comfortable being specific here that the operator is 'BEFORE' in case we add more operators however unlikey.

File: src/conditionalLogicService/generateFormElementsConditionallyShown.ts [hunk 7]
```diff
@@ -319,22 +349,79 @@ const generateFormElementsConditionallyShownWithParent = ({
...
  * ```
  *
  * @param parameters
  * @returns
  */
-export function generateFormElementsConditionallyShown(
-  parameters: FormElementsConditionallyShownParameters,
-): FormElementsConditionallyShown {
-  return generateFormElementsConditionallyShownWithParent(parameters)
+export function generateFormElementsConditionallyShown({
```
Comment (divporter) [inline src/conditionalLogicService/generateFormElementsConditionallyShown.ts:369]: So I can understand, why does this function now have all this logic following it, where previously it just called `generateFormElementsConditionallyShownWithParent`?

Comment (mymattcarroll) [reply src/conditionalLogicService/generateFormElementsConditionallyShown.ts:369]: 2 reasons
1. simplification + performance of apps-react (and other calling code) for working out elements vs submission conditional logic by not requiring full iteration of element tree again
2. this plays nicer with the types now that evaluateConditionalPredicates() accepts submission timestamp predicate but enabledSubmission does not support it.

File: src/conditionalLogicService/evaluateConditionalSubmissionTimestampPredicate.ts (hunk 6 above)
Comment (mymattcarroll) [reply src/conditionalLogicService/evaluateConditionalSubmissionTimestampPredicate.ts]: 👍

File: src/conditionalLogicService/generateFormElementsConditionallyShown.ts (hunk 7 above)
Comment (divporter) [reply src/conditionalLogicService/generateFormElementsConditionallyShown.ts:369]: Right, so an accompanying change in apps-react. On board, thank you.

File: src/conditionalLogicService/evaluateConditionalSubmissionTimestampPredicate.ts (hunk 5 above)
Comment (mymattcarroll) [reply src/conditionalLogicService/evaluateConditionalSubmissionTimestampPredicate.ts]: Fixed.

File: src/conditionalLogicService/evaluateConditionalSubmissionTimestampPredicate.ts (hunk 6 above)
Comment (mymattcarroll) [reply src/conditionalLogicService/evaluateConditionalSubmissionTimestampPredicate.ts]: Fixed

File: src/conditionalLogicService/generateFormElementsConditionallyShown.ts [hunk 8]
```diff
@@ -319,22 +349,79 @@ const generateFormElementsConditionallyShownWithParent = ({
...
+        formElementsCtrl,
+      })
+      if (!predicateElement) {
+        return false
+      }
+
+      // Reuse visibility already computed above instead of re-evaluating
+      // conditionally show predicates for the predicate element / parents.
+      const shown = formElementsConditionallyShown[predicateElement.name]
+      return !!shown && !shown.isHidden
```
[the remark this answers is not in the document]
Comment (mymattcarroll) [reply src/conditionalLogicService/generateFormElementsConditionallyShown.ts:412]: You are incorrect, I have added unit tests to ensure it works as expected.
